File: api/rust/cargo/lief/src/objc.rs
//! Module for Objective-C metadata
//!
//! This module contains the support of Objective-C metadata embedded in Mach-O
//! [`crate::macho::Binary`].
//!
//! One can access this metadata using the [`crate::macho::Binary::objc_metadata`] function
//!
//! ```
//! fn inspect_objc(macho: &lief::macho::Binary) {
//!     let metadata: lief::objc::Metadata = macho.objc_metadata().expect("Can't access metadata");
//!     for class in metadata.classes() {
//!         println!("{}", class.to_decl());
//!     }
//! }
//! ```

pub mod metadata;
pub mod class;
pub mod protocol;
pub mod method;
pub mod property;
pub mod ivar;
pub mod decl_opt;

#[doc(inline)]
pub use metadata::Metadata;

#[doc(inline)]
pub use class::Class;

#[doc(inline)]
pub use method::Method;

#[doc(inline)]
pub use property::Property;

#[doc(inline)]
pub use protocol::Protocol;

#[doc(inline)]
pub use ivar::IVar;

#[doc(inline)]
pub use decl_opt::DeclOpt;

File: doc/code/python/objc.py
import lief


def check(macho: lief.MachO.Binary) -> None:
    # lief-doc: check-start
    macho: lief.MachO.Binary

    metadata = macho.objc_metadata
    if metadata is not None:
        print("Objective-C metadata found")
    # lief-doc: check-end


def inspect(macho: lief.MachO.Binary) -> None:
    # lief-doc: inspect-start
    macho: lief.MachO.Binary

    metadata = macho.objc_metadata
    for clazz in metadata.classes:
        print(f"name={clazz.name}")
        for meth in clazz.methods:
            print(f"  method.name={meth.name}")
    print(metadata.to_decl())
    # lief-doc: inspect-end


# lief-doc: no-address-start
def print_without_address(macho: lief.MachO.Binary):

    metadata = macho.objc_metadata

    config = lief.objc.DeclOpt()
    config.show_annotations = False

    for cls in metadata.classes:
        print(cls.to_decl(config))


# lief-doc: no-address-end
